### PR TITLE
[Form] Added an AbstractChoiceLoader

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | ~   
| License       | MIT
| Doc PR        | TODO

Follow up #30962 and #30966.

~Based on #30966 , only review lat commit until it is merged.~

#EUFOSSA

### CODE REVIEW

**Review by vudaltsov — APPROVED** [review]
Looks great!

File: src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public function loadValuesForChoices(array $choices, $value = null)
+    {
+        // Optimize
+        if (empty($choices)) {
+            return [];
+        }
+
+        if ($value) {
+            // if a value callback exists, use it
+            return array_map($value, $choices);
```
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php]: This is actually a new optimisation that will make some good when using `choice_value` option :)

File: src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return $this->doLoadChoicesForValues($values, $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadValuesForChoices(array $choices, $value = null)
+    {
+        // Optimize
+        if (empty($choices)) {
```
Comment (fabpot) [inline src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php]: Some implementations use `$values = array_filter($values);` before the test. Is it something we can/want to add here?

File: src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return $this->choiceList = new ArrayChoiceList($this->loadChoices(), $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoicesForValues(array $values, $value = null)
+    {
+        // Optimize
+        if (empty($values)) {
```
Comment (fabpot) [inline src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php]: `if (!$values)`?

File: src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php (hunk 2 above)
Comment (HeahDude) [reply src/Symfony/Component/Form/ChoiceList/Loader/AbstractChoiceLoader.php]: It happens that an empty value may be a valid choice for any choice list, but we know it's not the case with intl base choice types, this is why we optimized their loaders specifically.

**Comment by fabpot** [comment]
@HeahDude What's the status of this PR? Are you willing to finish it?

**Comment by yceruto** [comment]
Taking over in https://github.com/symfony/symfony/pull/33218

**Pull request closed without merging**